sdk/python/tests/local_runner_test.py

```diff
@@ -0,0 +1,213 @@
+
+    with open(dst, "w") as f:
+        json.dump(["hello", "world", "kfp"], f)
+
+
+@light_component()
+def component_connect_demo(
+    src: kfp.components.InputPath(), dst: kfp.components.OutputPath()
+):
+    with open(src, "r") as f:
+        line = f.readline()
+        print(f"read first line: {line}")
+        with open(dst, "w") as fw:
+            fw.write(f"{line} copied")
+
+
+class LocalRunnerTest(unittest.TestCase):
+    def setUp(self):
+        import tempfile
+
+        self.temp_file = tempfile.NamedTemporaryFile()
+        with open(self.temp_file.name, "w") as f:
+            f.write("hello world")
+
+    def test_run_local(self):
+        def _pipeline(name: str):
+            hello(name)
+
+        run_pipeline_func_locally(
+            _pipeline,
+            {"name": "world"},
+            execution_mode=LocalClient.ExecutionMode("local"),
+        )
+
+    def test_local_file(self):
+        def _pipeline(file_path: str):
+            local_loader(file_path)
+
+        run_result = run_pipeline_func_locally(
+            _pipeline,
+            {"file_path": self.temp_file.name},
+            execution_mode=LocalClient.ExecutionMode("local"),
+        )
+        output_file_path = run_result.get_output_file("local-loader")
+
+        with open(output_file_path, "r") as f:
+            line = f.readline()
+            assert "hello" in line
+
+    def test_condition(self):
+        def _pipeline():
+            _flip = flip_coin()
+            with kfp.dsl.Condition(_flip.output == "head"):
+                hello("head")
+
+            with kfp.dsl.Condition(_flip.output == "tail"):
+                hello("tail")
+
+        run_pipeline_func_locally(
+            _pipeline, {}, execution_mode=LocalClient.ExecutionMode("local")
+        )
+
+    def test_for(self):
+        @light_component()
+        def cat(item, dst: OutputPath):
+            with open(dst, "w") as f:
+                f.write(item)
+
+        def _pipeline():
+            with kfp.dsl.ParallelFor(list().output) as item:
+                cat(item)
+
+        run_pipeline_func_locally(
+            _pipeline, {}, execution_mode=LocalClient.ExecutionMode("local")
+        )
+
+    def test_connect(self):
+        def _pipeline():
+            _local_loader = local_loader(self.temp_file.name)
+            component_connect_demo(_local_loader.output)
+
+        run_result = run_pipeline_func_locally(
+            _pipeline, {}, execution_mode=LocalClient.ExecutionMode("local")
+        )
+        output_file_path = run_result.get_output_file("component-connect-demo")
+
+        with open(output_file_path, "r") as f:
+            line = f.readline()
+            assert "copied" in line
+
+    def test_command_argument_in_any_format(self):
+        def echo():
+            return kfp.dsl.ContainerOp(
+                name="echo",
+                image=BASE_IMAGE,
+                command=["echo", "hello world", ">", "/tmp/outputs/output_file"],
```

Comment:
Are you sure this example works? The `>` file redirection requires a shell, but this command line does not use one.

Reply:
@Ark-kun,  yes, it works,  since the command will be executed through the shell.
```
def _run_group_dag(
    ...
                process = subprocess.Popen(
                    ...
                    shell="sh" != cmd[0],
                    ...
                )
```